src/server/mod.rs

```diff
@@ -0,0 +1,74 @@
+use anyhow::{Context, Result};
+use bb8_redis::{
+    bb8::{Pool, PooledConnection},
+    redis::{AsyncCommands, RedisResult},
+    RedisConnectionManager,
+};
+use std::{collections::HashMap, hash::Hash};
+
+use crate::msg::Message;
+
+pub struct Server<C, F>
+where
+    C: ToString,
+    F: Fn(String),
+{
+    pool: Pool<RedisConnectionManager>,
+    channels: HashMap<C, F>,
+}
+
+impl<C, F> Server<C, F>
+where
+    C: ToString + Eq + Hash,
+    F: Fn(String),
+{
+    pub fn new(pool: Pool<RedisConnectionManager>) -> Self {
+        let channels = HashMap::new();
+        Self { pool, channels }
+    }
+
+    async fn get_connection(&self) -> Result<PooledConnection<'_, RedisConnectionManager>> {
+        let conn = self
+            .pool
+            .get()
+            .await
+            .context("unable to retrieve a redis connection from the pool")?;
+
+        Ok(conn)
+    }
+
+    pub fn add_channel(mut self, cmd: C, handler: F) -> Self {
+        self.channels.insert(cmd, handler);
+        self
+    }
+
+    pub async fn exec(&self) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        loop {
+            for (cmd, handler) in self.channels.iter() {
+                let msg_res: RedisResult<Message> =
+                    conn.lpop(format!("msgbus.{}", cmd.to_string()), None).await;
+
```

Comment:
This is wrong. you will end up in a busy loop. Please try to find an optimal solution where u rely on the features  of the services you are using. 
- lpop is none blocking it means u will keep getting nulls until there is actually a message on (one of the queues) this is waste of CPU cycles.
- Hence again it's better for this use case to use a `brpop`
- brpop accepts `multiple` queues to wait on at the same time so you can do `brpop queu1 queue2 queue3 ...` etc.
- When means you can create a single`brpop` command and wait forever for ANY command available on any of the queueus
- once a mesasge arrives, you can tell from the return value of the brpop (it always return `(queue, data)` which queue returned a value
- Route your request to proper command.